target/linux/ath79/image/common-nec.mk

```diff
@@ -0,0 +1,75 @@
+DEVICE_VARS += NEC_FW_TYPE
+
+define Build/nec-append-uboot
+  cat $(STAGING_DIR_IMAGE)/$(SOC)_nec_aterm-u-boot.bin >> $@
+endef
+
+define Build/nec-boot-header
+  $(eval flags=$(if $(1),$(word 1,$(1)),0x0002))
+  $(eval addr=$(if $(word 2,$(1)),$(word 2,$(1)),80060000))
+  $(eval databin=$(if $(word 3,$(1)),$(word 3,$(1)),$@))
+  ( \
+    flags=$$(printf "%04x%04x" $(flags) $$((~$(flags) & 0xffff))); \
+    nec_fw_size=$$(printf '%08x' "$$(($$(stat -c%s $(databin)) + 0x18))"); \
+    printf $$(echo "$${flags}$${nec_fw_size}0000001800000000$(addr)$(addr)" | \
+      sed 's/../\\x&/g'); \
+    cat $(databin); \
+  ) > $(databin).new
+  mv $(databin).new $(databin)
+  ( \
+    cksum=$$( \
+      dd if=$(databin) ibs=4 skip=1 | od -A n -t u2 --endian=little | \
+        tr -s ' ' '\n' | \
+        awk '{s+=$$0}END{printf "%04x", 0xffff-(s%0x100000000)%0xffff}'); \
+    printf "\x$${cksum:2:2}\x$${cksum:0:2}" | \
+      dd of=$(databin) conv=notrunc bs=2 seek=6 count=1; \
+  )
+endef
+
+define Build/nec-usbaterm-fw
+  ( \
+    printf "VERSION: 9.99.99\nOEM1 VERSION: 9.9.99\n"; \
+    printf "$(call toupper,$(LINUX_KARCH)) $(VERSION_DIST) Linux-$(LINUX_VERSION)\n"; \
+  ) | dd of=$@.copyblk bs=4 conv=sync
+  touch $@.endblk
+  $(call Build/nec-boot-header,0x0000 00000000 $@.copyblk)
+  $(call Build/nec-boot-header,0x0001 00000000 $@.endblk)
+  ( \
+    printf "USB ATERMWL3050\x00"; \
+    printf $$(echo "ffffffffffffffffffffffffffffffff" | sed 's/../\\x&/g'); \
+    cat $@.copyblk $@ "$(KDIR)/loader-$(DEVICE_NAME).bin" $@.endblk; \
+  ) > $@.new
+  printf "Binary Type$(1) File END \r\n" > $@.footer
+  ( \
+    pad=$$((0x20 - $$(stat -c%s $@.footer))); \
+    for i in $$(seq 1 $$pad); do printf "\xff"; done; \
+    cat $@.footer; \
+  ) >> $@.new
+  mv $@.new $@
+  rm -f $@.copyblk $@.endblk $@.footer
+endef
+
+define Build/remove-uimage-header
+  dd if=$@ of=$@.new iflag=skip_bytes skip=64 2>/dev/null
+  mv $@.new $@
+endef
+
+define Device/nec-netbsd-aterm
+  DEVICE_VENDOR := NEC
+  LOADER_TYPE := bin
+  KERNEL := kernel-bin | append-dtb | lzma | loader-kernel | uImage none
```

Comment:
Considering that the bootloader has been replaced, do we still need a lzma kernel loader for KERNEL?

Reply:
Yes, it's needed to decompress the kernel since the U-Boot with the current configuration doesn't support decompression with any methods.
On NetBSD-based Aterm devices, only `0x20000` (128KiB) is available for a bootloader on the spi-nor flash and that limitation is too small for mainline U-Boot with the default options. So many features/commands not required for booting and recoverying are disabled on that devices.

example:

- networking support
- FIT support
- all decompression method support

etc...

compiled U-Boot binary with the current changes:

`-rw-r--r-- 1 tofu tofu  130788 May 12 17:59 openwrt-ath79-tiny-nec_wg600hp-uboot.bin`

available commands:

```
=>  help
?         - alias for 'help'
boot      - boot default, i.e., run 'bootcmd'
bootd     - boot default, i.e., run 'bootcmd'
bootm     - boot application image from memory
clocks    - display clocks
env       - environment handling commands
help      - print command description/usage
loadb     - load binary file over serial line (kermit mode)
loadx     - load binary file over serial line (xmodem mode)
loady     - load binary file over serial line (ymodem mode)
panic     - Panic with optional message
printenv  - print environment variables
reset     - Perform RESET of the CPU
setenv    - set environment variables
version   - print monitor, compiler and linker version
```